analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Extensions/IOperationExtensions.cs

```diff
@@ -22,17 +22,27 @@ namespace SonarAnalyzer.SymbolicExecution.Roslyn
 {
     internal static class IOperationExtensions
     {
-        internal static ISymbol TrackedSymbol(this IOperation operation) =>
-            operation?.Kind switch
+        internal static ISymbol TrackedSymbol(this IOperation operation)
+        {
+            return operation?.Kind switch
             {
-                OperationKindEx.Conversion => TrackedSymbol(operation.ToConversion().Operand),
-                OperationKindEx.FieldReference when IFieldReferenceOperationWrapper.FromOperation(operation) is var fieldReference && IsStaticOrThis(fieldReference) => fieldReference.Field,
-                OperationKindEx.LocalReference => ILocalReferenceOperationWrapper.FromOperation(operation).Local,
-                OperationKindEx.ParameterReference => IParameterReferenceOperationWrapper.FromOperation(operation).Parameter,
-                OperationKindEx.Argument => IArgumentOperationWrapper.FromOperation(operation).Value.TrackedSymbol(),
+                OperationKindEx.Conversion when operation.ToConversion() is var conversion && !IsTryDownCast(conversion) => TrackedSymbol(conversion.Operand),
+                OperationKindEx.FieldReference when operation.ToFieldReference() is var fieldReference && IsStaticOrThis(fieldReference) => fieldReference.Field,
+                OperationKindEx.LocalReference => operation.ToLocalReference().Local,
+                OperationKindEx.ParameterReference => operation.ToParameterReference().Parameter,
+                OperationKindEx.Argument => operation.ToArgument().Value.TrackedSymbol(),
+                OperationKindEx.DeclarationExpression => IDeclarationExpressionOperationWrapper.FromOperation(operation).Expression.TrackedSymbol(),
+                OperationKindEx.PropertyReference when operation.ToPropertyReference() is { Property: { IsVirtual: false } property } propertyReference
```

Comment:
The pattern match filters and deconstructs into non-null variables.
There are no clear conventions here. I just personally like the expressiveness and null safety of patterns and I'd like to avoid the multiple dereferences of `propertyReference.Property` which would be needed without the deconstruction.

> I believe Property is always non-null

That is the problem. As long as we are not on a newer version of Roslyn and you we don't have nullable annotations, you just don't know. There are many broken code scenarios where you some essential properties are null (not in this case though).
I have an old PR where I updated the shimlayer generation to include the nullable annotations, documentation comments, and some nice extension methods here https://github.com/SonarSource/sonar-dotnet/pull/6078 I will open it again (it was closed due to time constraints) and see if we can get it merged.

Reply:
see https://github.com/SonarSource/sonar-dotnet/pull/7007#issuecomment-1497467154